### DISCUSSION TITLE
How to setup the parameters for different usecases?

### DISCUSSION THREAD

**Discussion by tallner** [bitcraze/discussions#294] [Q&A]
Are there any documentation for use cases when using the client libraries(Java in my case)?
It is sometimes a bit hard to know what the parameters are used for (in a use case), and sometimes I assume several parameters needs to be used or changed.
Would be awesome with a few examples on how to achieve a certain functionality and what parameters do I need to set in order to fulfil the use case.

In my case I am writing a JavaFX desktop client using the Javalib.
Just bought the Flow Deck V2 and the Multiranger and want to implement a hover mode and flight assistant mode as in the python client.
Later on I want to extend it to autonomous flight.

Is there any documentation how to get started or how to implement these functionalities?
I assume the hover mode in the Python client have some documentation?

Thanks
Chris

**Comment by krichardsson** [comment]
Yes, that would be nice, unfortunately there is no such documentation :-(
 
I think the best place to look is in the python client source code. It is slightly convoluted, but the choosing of mode is done in [FlightTab.py](https://github.com/bitcraze/crazyflie-clients-python/blob/master/src/cfclient/ui/tabs/FlightTab.py). Search for `_assist_mode_combo`.
The setting of this combo will change [the type of set point that is sent to the Crazyflie](https://github.com/bitcraze/crazyflie-clients-python/blob/53de36d344cd06ac73cc4093fd8f77d704501f73/src/cfclient/ui/main.py#L196-L209)

You will find the implementation of [the set point functions in the python lib](https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/commander.py)

Also note that some parameters are set automatically in the Crazyflie based on the decks that are mounted. In your case the flow deck will cause a switch to the kalman estimator (parameter `stabilizer.estimator`)

**Comment by fredg02** [comment]
Use-case specific documentation is indeed missing. 

So for your use case (Flow Deck and Multiranger) here are a few tips:
* First of all, your client should check if the required decks are connected. 
  * Check that param group `deck` and names `bcFlow` or `bcFlow2` and ` bcMultiranger ` exist (see https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#deck-pin-allocation (edited) for the full list of deck names)
* If the decks are found you should be able to use the [ZDistancePacket](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib/src/main/java/se/bitcraze/crazyflie/lib/crtp/ZDistancePacket.java) instead of the default [CommanderPacket](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib/src/main/java/se/bitcraze/crazyflie/lib/crtp/CommanderPacket.java)
  * With the [ZDistancePacket](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib/src/main/java/se/bitcraze/crazyflie/lib/crtp/ZDistancePacket.java) you can set a value for the distance to the ground.
  * You need to continuously use this packet in hover mode.
    * Please note that the distance sensors have some limits and you should stay within reasonable boundaries to not risk a crash due to a false reading.
* You should be able to test the Multiranger deck with the [LoggingMultirangerExample.java](https://github.com/fredg02/se.bitcraze.crazyflie.lib/blob/master/se.bitcraze.crazyflie.lib.examples/src/main/java/se/bitcraze/crazyflie/lib/examples/LoggingMultirangerExample.java). If that does not work, the logging group or names might have changed in recent firmwares.

**Reply by tallner** [reply]
Good evening!
I did not connect the Multiranger, only the Flow deck.
I tried using both ZDistancePacket and HoverPacket.
CF will hover at the height set, but it is drifting a bit, specially with ZDistancePackage.
I did not set any parameters, but I read them to make sure the deck is found and the kalman estimator is set.

The only code I actually wrote is this
this.mCrazyflie.sendPacket(new HoverPacket(0, 0, 0, 0.1f));
try {
    Thread.sleep(100);
} catch (InterruptedException e) {
    e.printStackTrace();
}

But when running the python client the CF is not drifting at all, it is very stable at the point is hovering above.

Could I be missing some setting?

**Reply by tallner** [reply]
Any thoughts on this?

**Reply by krichardsson** [reply]
This looks like a new question (even though it is related), please start a new discussion.

**Comment by knmcguire** [comment]
Just a side node, @fredg02 I changed one of the links you gave here since we are out phasing the wiki. It's all hosted on www.bitcraze.io now.
